### DISCUSSION TITLE
Ability to execute commands

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#12] [Ideas]
I'm looking to build in the ability to execute commands straight from the toolbar, without the need to use another plugin like Advanced URI (would still be an option if that's what you need).

This should make it easier for novice users, require the use of one less plugin, and allow for copy/pasting straight from the [examples](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Examples) in the User Guide.

Toolbar items would then allow for note names, URLs/URIs, and commands... with (hopefully) more options to come!

## User experience

- Ideally the Settings UI would allow for a slash `/` to be put at the front of links, which would trigger a "CommandSuggester" (ideally), or (really short term) would require the full command to be entered. Aiming for the former.
  - Update (compromise?) 2024-03-30: Now planning to have a switcher/dropdown where you pick the type of link.

  ![CleanShot 2024-03-28 at 11 35 51@2x](https://github.com/chrisgurney/obsidian-note-toolbar/assets/17337703/cf76b0b1-af18-40c8-abe6-611fa7f72a3b)

## Implementation notes

- Commander uses this API: `app.commands.executeCommandById('editor:focus')`
- We would first need to get the command ID, hence the probable need for a "CommandSuggester" to list out available commands, which I believe is possible via `Object.values(app.commands.commands);`

**Comment by FelipeRearden** [comment]
This is the templater script that I use to get the commandID from a suggester when I am building my Scripts

```
<%*
const cmd = await tp.system.suggester(cmd => cmd.name, Object.values(this.app.commands.commands), false, "𐙟", 10)
await navigator.clipboard.writeText(`app.commands.executeCommandById('${cmd.id}')`)
%>
```

Maybe could help 🙏

**Comment by chrisgurney** [comment]
Now available in the beta! [Check it out.](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.4.0)
